### PR TITLE
[issue-7357] Add support for Avro logical types in both realtime and batch ingestion

### LINKED ISSUE #7357: Support for Avro logical types in realtime tables
It seems like Avro logical types are not supported in the Avro message decoders.

I have been browsing through the code and checked the past issues and I have not found mentions to Avro logical types.

I will be happy to open a PR when I work out the best place and how to add these changes.

**More details about this change**:

Avro has support for logical types - more details at [Avro spec docs](https://avro.apache.org/docs/current/spec.html#Logical+Types). A subject schema can therefore be defined as 

```
{
...
    {
      "name": "amount",
      "type": {
        "logicalType": "decimal",
        "precision": 64,
        "scale": 2,
        "type": "bytes"
      }
    },
...
}
```
Pinot Avro decoder does not support logical types at the moment there for trying to represent this value a `String` column will result in a wrong representations of the value. Trying to represent this value as a `Float`/`Double` column will results in a exception:

_Table schema:_
```
{
  "schemaName": "some_table",
  ...
  "metricFieldSpecs": [
    ...
    {
      "name": "amount",
      "dataType": "DOUBLE",
      "defaultNullValue": 0
    },
    ...
  ],
  ...
}
```

_Exception:_
```
Caught exception while transforming the record: {
  "fieldToValueMap" : {
    ...
    "amount" : "abc=",
    ...
  },
  "nullValueFields" : [ ]
}
java.lang.RuntimeException: Caught exception while transforming data type for column: amount
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:95) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.segment.local.recordtransformer.CompositeTransformer.transform(CompositeTransformer.java:83) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.processStreamEvents(LLRealtimeSegmentDataManager.java:514) [pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.consumeLoop(LLRealtimeSegmentDataManager.java:417) [pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager$PartitionConsumer.run(LLRealtimeSegmentDataManager.java:564) [pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.UnsupportedOperationException: Cannot convert value from BYTES to DOUBLE
	at org.apache.pinot.common.utils.PinotDataType$12.toDouble(PinotDataType.java:617) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.common.utils.PinotDataType$8.convert(PinotDataType.java:429) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.common.utils.PinotDataType$8.convert(PinotDataType.java:386) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:90) ~[pinot-all-0.8.0-SNAPSHOT-jar-with-dependencies.jar:0.8.0-SNAPSHOT-bb34406aaa205f5c85b88c928d477fd267eda1b4]
	... 5 more

```

Ideally Pinot should be able to perform automatic logical type conversion if the type details are provided.

### PR DESCRIPTION
**Pull request by ddcprg**
## Description

Fixes #7357 

I'm looking to get some feedback before making this a full PR

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

The code change should be backwards compatible.

### CODE REVIEW

**Comment by ddcprg** [comment]
I'm looking at modifying the data used by `RealtimeClusterIntegrationTest` to add fields with logical types. Any guidelines on how this data has been generated?

**Comment by ddcprg** [comment]
I thought changing the commit hash would re-trigger the build

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    synchronized (CONVERSION_MAP) {
+      CONVERSION_MAP.put(DECIMAL, new Conversions.DecimalConversion());
+      CONVERSION_MAP.put(UUID, new Conversions.UUIDConversion());
+      CONVERSION_MAP.put(DATE, new TimeConversions.DateConversion());
+      CONVERSION_MAP.put(TIME_MILLIS, new TimeConversions.TimeMillisConversion());
+      CONVERSION_MAP.put(TIME_MICROS, new TimeConversions.TimeMicrosConversion());
+      CONVERSION_MAP.put(TIMESTAMP_MILLIS, new TimeConversions.TimestampMillisConversion());
+      CONVERSION_MAP.put(TIMESTAMP_MICROS, new TimeConversions.TimestampMicrosConversion());
+    }
+  }
```
Comment (richardstartin) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java:57]: This doesn't need to be synchronised because static initialisers can only be invoked once per class loader, and if you look at the bytecode you'll notice a synthetic method called `<clinit>` in which `CONVERSION_MAP` is initialised, along with the block of code here.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+   * These constants are copied from org.apache.avro.LogicalTypes
+   */
+  private static final String DECIMAL = "decimal";
+  private static final String UUID = "uuid";
+  private static final String DATE = "date";
+  private static final String TIME_MILLIS = "time-millis";
+  private static final String TIME_MICROS = "time-micros";
+  private static final String TIMESTAMP_MILLIS = "timestamp-millis";
+  private static final String TIMESTAMP_MICROS = "timestamp-micros";
+  private static final Map<String, Conversion<?>> CONVERSION_MAP = new HashMap<>();
```
Comment (walterddr) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java:38]: can we put these in ArvoSchemUtils? is there a reason we need the Registry?

**Review by walterddr — COMMENTED** [review]
the approach looks good. thanks @ddcprg for the contribution. I left one minor comment otherwise good to go.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java (hunk 2 above)
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java:38]: I'll rebase, review the code since it's been a while and get back to you with the reason why I put this here

Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java:38]: there is not reason to not to move it `ArvoSchemUtils` other than having some separation of concern,, I'll move them over if you think that's a better place for this

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 3]
```diff
@@ -31,22 +34,22 @@
...
 
-
-/**
- * Extractor for Avro Records
- */
+/** Extractor for Avro Records */
 public class AvroRecordExtractor extends BaseRecordExtractor<GenericRecord> {
   private Set<String> _fields;
   private boolean _extractAll = false;
+  private boolean _applyLogicalTypes = false;
```
Comment (ddcprg) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:41]: this setting should come from table config in order to potentially break existing ingestion that come with logical types, however `RecordExtractorConfig` is always null. Is anyone working on this logic at the moment? I could give that a go as I'll need to add the setting for this feature

Comment (walterddr) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:41]: yes I think we can create a `AvroRecordExtractorConfig` and provide some defaults to this if necessary. see `CsvRecordExtractorConfig` as an example. i think it can be part of this PR

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java (hunk 2 above)
Comment (walterddr) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroLogicalTypeConversionRegistry.java]: this looks good now.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderConfig.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.apache.pinot.spi.data.readers.RecordReaderConfig;
+
+/**
+ * Config for {@link AvroRecordReader}
+ */
+public class AvroRecordReaderConfig implements RecordReaderConfig {
+  private boolean _enableLogicalTypes;
+
+  public boolean isEnableLogicalTypes() {
```
Comment (ddcprg) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderConfig.java:29]: For ingestion jobs this properties can be configured via recordsReaderSpec. However I can't see how these property are configured for a realtime table - perhaps via ingestionSpec in the table definition? I was looking through the code but I couldn't figure it out. I believe this is not possible at the moment

Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderConfig.java:29]: I think I've found where the properties can be read from. Stream config has decoder properties (it looks like these are available in table config of realtime tables) which are propagated to the Avro decoder classes, it looks like I can take advantage of that to populate the logical type flag. I'll add a commit to the PR and let me know if you see anything missing @walterddr

Comment (walterddr) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderConfig.java:29]: hmm. not sure which one are you referring to. there's a `RecordReaderConfig` base class which provides configuration properties to `RecordReader`s (see `RecordReaderFactory` class). and it is apparently passed in via a `readerConfigFile` path (see. `CreateSegmentCommand` entrypoint)

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorConfig.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Config for {@link AvroRecordExtractor}
+ */
+public class AvroRecordExtractorConfig implements RecordExtractorConfig {
+  private boolean _enableLogicalTypes = false;
+
+  @Override
+  public void init(Map<String, String> props) {
+    _enableLogicalTypes = Boolean.valueOf(props.getOrDefault("enableLogicalTypes", Boolean.FALSE.toString()));
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorConfig.java:33]:
```suggestion
    _enableLogicalTypes = Boolean.parseBoolean(props.get("enableLogicalTypes"));
```

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 6]
```diff
@@ -107,4 +120,23 @@ protected Object convertRecord(Object value) {
...
+  protected Object convertSingleValue(Object value) {
+    if (value instanceof BigDecimal) {
+      return Double.valueOf(((BigDecimal) value).doubleValue());
+    }
+    if (value instanceof LocalDate) {
+      return Long.valueOf(((LocalDate) value).toEpochDay());
+    }
+    if (value instanceof Instant) {
+      return Long.valueOf(((Instant) value).toEpochMilli());
+    }
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:138]: (minor) Can be simplified
```suggestion
    if (value instanceof BigDecimal) {
      return ((BigDecimal) value).doubleValue();
    }
    if (value instanceof LocalDate) {
      return ((LocalDate) value).toEpochDay();
    }
    if (value instanceof Instant) {
      return ((Instant) value).toEpochMilli();
    }
```

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 7]
```diff
@@ -20,19 +20,49 @@
...
+    CONVERSION_MAP.put(DECIMAL, new Conversions.DecimalConversion());
+    CONVERSION_MAP.put(UUID, new Conversions.UUIDConversion());
+    CONVERSION_MAP.put(DATE, new TimeConversions.DateConversion());
+    CONVERSION_MAP.put(TIME_MILLIS, new TimeConversions.TimeMillisConversion());
+    CONVERSION_MAP.put(TIME_MICROS, new TimeConversions.TimeMicrosConversion());
+    CONVERSION_MAP.put(TIMESTAMP_MILLIS, new TimeConversions.TimestampMillisConversion());
+    CONVERSION_MAP.put(TIMESTAMP_MICROS, new TimeConversions.TimestampMicrosConversion());
+  }
+
+  /** Returns the data type stored in Pinot that is associated with the given Avro type. */
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:65]: (format) We usually keep the javadoc multi lines

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 8]
```diff
@@ -61,9 +91,7 @@ public static DataType valueOf(Schema.Type avroType) {
     }
   }
 
-  /**
-   * @return if the given avro type is a primitive type.
-   */
+  /** @return if the given avro type is a primitive type. */
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:94]: (format) We usually keep the javadoc multi lines

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderTest.java [hunk 9]
```diff
@@ -39,7 +39,7 @@ public class AvroRecordReaderTest extends AbstractRecordReaderTest {
   protected RecordReader createRecordReader()
       throws Exception {
     AvroRecordReader avroRecordReader = new AvroRecordReader();
-    avroRecordReader.init(_dataFile, _sourceFields, null);
+    avroRecordReader.init(_dataFile, _sourceFields, new AvroRecordReaderConfig());
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordReaderTest.java:42]: We don't need to change this (also want to test the behavior of passing `null`)

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordToPinotRowGeneratorTest.java [hunk 10]
```diff
@@ -44,7 +44,7 @@ public void testIncomingTimeColumn()
     Set<String> sourceFields = Sets.newHashSet("incomingTime", "outgoingTime");
 
     AvroRecordExtractor avroRecordExtractor = new AvroRecordExtractor();
-    avroRecordExtractor.init(sourceFields, null);
+    avroRecordExtractor.init(sourceFields, new AvroRecordExtractorConfig());
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordToPinotRowGeneratorTest.java:47]: We don't need to change this (also want to test the behavior of passing `null`)

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java [hunk 11]
```diff
@@ -94,8 +95,10 @@ public void init(Map<String, String> props, Set<String> fieldsToRead, String top
     _deserializer = new KafkaAvroDeserializer(schemaRegistryClient);
     Preconditions.checkNotNull(topicName, "Topic must be provided");
     _topicName = topicName;
+    AvroRecordExtractorConfig config = PluginManager.get().createInstance(AvroRecordExtractorConfig.class.getName());
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java:98]:
```suggestion
    AvroRecordExtractorConfig config =  new AvroRecordExtractorConfig();
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractorConfig.java [hunk 12]
```diff
@@ -18,8 +18,13 @@
...
 
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+
 /**
  * Interface for configs of {@link RecordExtractor}
  */
 public interface RecordExtractorConfig {
+  default void init(@Nonnull Map<String, String> props) { }
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractorConfig.java:29]: (convention) We don't usually annotate `non-null`, but only annotate `nullable`. Also make it multiple lines
```suggestion
  default void init(Map<String, String> props) {
  }
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise, no major comments

Does the change also apply to batch ingestion flow? If so, let's update the PR name and description to reflect that

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 13]
```diff
@@ -31,16 +34,19 @@
 import org.apache.pinot.spi.data.readers.GenericRow;
 import org.apache.pinot.spi.data.readers.RecordExtractorConfig;
 
-
-/**
- * Extractor for Avro Records
- */
+/** Extractor for Avro Records */
```
[the remark this answers is not in the document]
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:37]: not sure, perhaps an old pinot code style I had in my IDE

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 14]
```diff
@@ -107,4 +107,23 @@ protected Object convertRecord(Object value) {
...
   }
+
+  /**
+   * This method convert any Avro logical-type converted (or not) value to a class supported by
+   * Pinot {@link GenericRow}
+   */
+  @Override
+  protected Object convertSingleValue(Object value) {
+    if (value instanceof BigDecimal) {
+      return Double.valueOf(((BigDecimal) value).doubleValue());
```
[the remark this answers is not in the document]
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java]: once bigdecimal type is fully supported a record reader option can control this behaviour

**Comment by ddcprg** [comment]
@Jackie-Jiang thanks for reviewing, there is one more thing I'd like to do on this PR before I merge it. I'm looking at adding a full integration test for both batch and ingestion, if you have any pointer please let me know

**Comment by Jackie-Jiang** [comment]
@ddcprg You may take a look at `HybridClusterIntegrationTest` on how to feed the data into both batch flow and streaming flow. In order to add an integration test, we need to add a new avro data set with the logical types. IMO it might be too much for this support (adding an integration test can add 1-2 minutes to the CI). A thorough unit test on avro record extractor should be good enough.

**Comment by ddcprg** [comment]
Thanks @Jackie-Jiang in that case I'll add more unit tests around record extractor. I'll squash all the commits into one after that

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 15]
```diff
@@ -107,4 +121,25 @@ protected Object convertRecord(Object value) {
...
+   * represented at all.
+   * Similarly, timestamp microsecond precision is not supported at the moment. These values will get converted to
+   * millisecond precision.
+   */
+  @Override
+  protected Object convertSingleValue(Object value) {
+    if (value instanceof BigDecimal) {
+      return ((BigDecimal) value).doubleValue();
+    }
+    if (value instanceof Instant) {
```
Comment (ddcprg) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:139]: I'm thinking it may be better to initially return a String representation of the timestamp instead, Pinot can coerce the value into a timestamp. At this point I can't make a guess of whether the user has specified millis or micros. Behaviour control could be configured via record reader settings in the future. @Jackie-Jiang  @walterddr let me know what you think

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 16]
```diff
@@ -33,14 +35,19 @@
 
 
 /**
- * Extractor for Avro Records
+ *  Extractor for Avro Records
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:38]: (minor) revert

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 17]
```diff
@@ -20,18 +20,41 @@
...
   private AvroSchemaUtil() {
   }
 
+  public static Conversion<?> findConversionFor(String typeName) {
+    return CONVERSION_MAP.get(typeName);
+  }
+
   /**
-   * Returns the data type stored in Pinot that is associated with the given Avro type.
+   *  Returns the data type stored in Pinot that is associated with the given Avro type.
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:57]: (minor) Revert, same for other places (is this auto-formatted?)

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 18]
```diff
@@ -114,4 +137,39 @@ private static ArrayNode convertStringsToJsonArray(String... strings) {
...
+      return value;
+    }
+    Conversion<?> conversion = AvroSchemaUtil.findConversionFor(logicalType.getName());
+    if (conversion == null) {
+      return value;
+    }
+    return Conversions.convertToLogicalType(value, field.schema(), logicalType, conversion);
+  }
+
+  static {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:166]: (minor) Let's put this static block next to the map construction for readability

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java (hunk 15 above)
Comment (Jackie-Jiang) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:139]: Seems the `Instant` is always converted from the timestamp type, so we might want to return the `Timestamp` object here.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java (hunk 17 above)
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:57]: yes, not sure why my IDE is doing this

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java (hunk 15 above)
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java:139]: this now returns `java.sql.Timestamp` instead

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java [hunk 19]
```diff
@@ -107,4 +122,25 @@ protected Object convertRecord(Object value) {
...
+   * Pinot {@link GenericRow}
+   *
+   * Note that at the moment BigDecimal is converted to Pinot double which may lead to precision loss or may not be
+   * represented at all.
+   * Similarly, timestamp microsecond precision is not supported at the moment. These values will get converted to
+   * millisecond precision.
+   */
+  @Override
+  protected Object convertSingleValue(Object value) {
+    if (value instanceof BigDecimal) {
```
[the remark this answers is not in the document]
Comment (ddcprg) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractor.java]: this is great! I've now changed this line and squashed all the commits into a single one
